src/backend/distributed/commands/table.c

```diff
@@ -820,23 +825,39 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand,
 		else if (alterTableType == AT_AttachPartition)
 		{
 			PartitionCmd *partitionCommand = (PartitionCmd *) command->def;
+			Oid attachedRelationId = RangeVarGetRelid(partitionCommand->name, NoLock,
+													  false);
+			char attachedRelationKind = get_rel_relkind(attachedRelationId);
 
 			/*
-			 * We only support ALTER TABLE ATTACH PARTITION, if it is only subcommand of
-			 * ALTER TABLE. It was already checked in ErrorIfUnsupportedAlterTableStmt.
+			 * We support ALTER INDEX ATTACH PARTITION and ALTER TABLE ATTACH PARTITION
+			 * if it is only subcommand of ALTER TABLE command. Since the attached relation
+			 * type is index for ALTER INDEX ATTACH PARTITION, we need to use the relation
+			 * id this index is created for.
+			 *
+			 * Both were already checked in ErrorIfUnsupportedAlterIndexStmt and
+			 * ErrorIfUnsupportedAlterTableStmt.
 			 */
-			Assert(list_length(commandList) <= 1);
+			if (attachedRelationKind == RELKIND_INDEX)
+			{
+				bool missingOk = false;
+				rightRelationId = IndexGetRelation(attachedRelationId, missingOk);
```

Comment:
Do we need to check if `attachedRelationId` `IsCitusTable` ?

Reply:
Since partition must be the partition of left table which has already checked above to be Citus table, right relation must be Citus table. I've added an assertion.